include/aws/common/priority_queue.h

```diff
@@ -0,0 +1,88 @@
+#ifndef AWS_PRIORITY_QUEUE_H
+#define AWS_PRIORITY_QUEUE_H
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/array_list.h>
+
+struct aws_priority_queue {
+    int(*compare)(void *a, void *b);
```

Comment:
Do we need to pass a context pointer? How would one implement a compare function that, say, calls back into java?

Reply:
It seems to me, that you'd just pass a function pointer, where the implementation knows to callback into java.

Reply:
Ah, but if you have multiple compare functions on the java side, you only have one trampoline (unless you do runtime code generation). How does the trampoline know which java-side function _object_ to call, for instance?